test/loading.jl

```diff
@@ -1274,3 +1274,60 @@ end
         @test_throws ErrorException Base.check_src_module_wrap(p, fpath)
     end
 end
+
+@testset "relocatable upgrades #51989" begin
+    mktempdir() do depot
+        # Create fake `Foo.jl` package with two files:
+        foo_path = joinpath(depot, "dev", "Foo")
+        mkpath(joinpath(foo_path, "src"))
+        open(joinpath(foo_path, "src", "Foo.jl"); write=true) do io
+            println(io, """
+            module Foo
+            include("internal.jl")
+            end
+            """)
+        end
+        open(joinpath(foo_path, "src", "internal.jl"); write=true) do io
+            println(io, "const a = \"asd\"")
+        end
+        open(joinpath(foo_path, "Project.toml"); write=true) do io
+            println(io, """
+            name = "Foo"
+            uuid = "00000000-0000-0000-0000-000000000001"
+            version = "1.0.0"
+            """)
+        end
+
+        # In our depot, `dev` and then `precompile` this `Foo` package.
+        @test success(addenv(
+            `$(Base.julia_cmd()) --startup-file=no -e 'import Pkg; Pkg.develop("Foo"); Pkg.precompile(); exit(0)'`,
+            "JULIA_DEPOT_PATH" => depot,
+        ))
+
+        # Get the size of the generated `.ji` file so that we can ensure that it gets altered
+        foo_compiled_path = joinpath(depot, "compiled", "v$(VERSION.major).$(VERSION.minor)", "Foo")
+        cache_path = joinpath(foo_compiled_path, only(filter(endswith(".ji"), readdir(foo_compiled_path))))
```

Comment:
Just a minor thing: One could push `depot` onto the LOAD_PATH and use something like (untested)
```
pkg = Base.identify_package("Foo")
Base.compilecache(pkg)
```
to get all `.ji` files. Same can be done below.

Reply:
I didn't feel this was important enough to justify waiting another 24 hours for CI to catch up; if you feel strongly about it, feel free to open another PR.